flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/autoscaler/ConfigMapStore.java

```diff
@@ -67,19 +67,14 @@ protected Optional<String> getSerializedState(
     }
 
     protected void removeSerializedState(KubernetesJobAutoScalerContext jobContext, String key) {
-        getConfigMap(jobContext)
-                .ifPresentOrElse(
-                        configMap -> configMap.getData().remove(key),
-                        () -> {
-                            throw new IllegalStateException(
-                                    "The configMap isn't created, so the remove is unavailable.");
-                        });
+        getConfigMap(jobContext).ifPresent(configMap -> configMap.getData().remove(key));
     }
 
     public void flush(KubernetesJobAutoScalerContext jobContext) {
         Optional<ConfigMap> configMapOpt = cache.get(jobContext.getJobKey());
         if (configMapOpt == null || configMapOpt.isEmpty()) {
             LOG.debug("The configMap isn't updated, so skip the flush.");
+            // Do not flush if there are no updates.
```

Comment:
I'm thinking whether any updating or removeing can be determined, please give me a few times to test it.

Reply:
I write 2 tests:
1. one is `testRemoveAndFlushEveryTime`, it's similar to current PR code
2. one is `testQueryAndFlushIfNeeded`, it's similar to master code

- `testRemoveAndFlushEveryTime` will remove and flush every time, and it will call kubernetes every time.
- `testQueryAndFlushIfNeeded` will query from cache, and value is null, so remove and flush isn't necessary. So it doens't call kubernetes.

My question is: the extra call is expected? And I think this PR is fine if it doesn't introuduce any extra opration.

These 2 tests can be added to ConfigMapStoreTest, and they can run directly.


```    
    @Test
    void testRemoveAndFlushEveryTime() {
        KubernetesJobAutoScalerContext ctx = createContext("cr1", kubernetesClient);
        var key = "key";
        var value = "value";

        var configMapStore = new ConfigMapStore(kubernetesClient);
        configMapStore.putSerializedState(ctx, key, value);
        assertEquals(2, mockWebServer.getRequestCount());
        configMapStore.flush(ctx);
        assertEquals(3, mockWebServer.getRequestCount());

        // Get from cache
        assertThat(configMapStore.getSerializedState(ctx, key)).hasValue(value);
        assertEquals(3, mockWebServer.getRequestCount());

        // Get from kubernetes
        configMapStore = new ConfigMapStore(kubernetesClient);
        assertThat(configMapStore.getSerializedState(ctx, key)).hasValue(value);
        assertEquals(4, mockWebServer.getRequestCount());

        // Remove the key
        configMapStore.removeSerializedState(ctx, key);
        assertEquals(4, mockWebServer.getRequestCount());

        configMapStore.flush(ctx);
        assertEquals(5, mockWebServer.getRequestCount());

        // Every flush will call kubernetes.
        configMapStore.flush(ctx);
        assertEquals(6, mockWebServer.getRequestCount());
        configMapStore.flush(ctx);
        assertEquals(7, mockWebServer.getRequestCount());
        configMapStore.flush(ctx);
        assertEquals(8, mockWebServer.getRequestCount());
    }


    @Test
    void testQueryAndFlushIfNeeded() {
        KubernetesJobAutoScalerContext ctx = createContext("cr1", kubernetesClient);
        var key = "key";
        var value = "value";

        var configMapStore = new ConfigMapStore(kubernetesClient);
        configMapStore.putSerializedState(ctx, key, value);
        assertEquals(2, mockWebServer.getRequestCount());
        configMapStore.flush(ctx);
        assertEquals(3, mockWebServer.getRequestCount());

        // Get from cache
        assertThat(configMapStore.getSerializedState(ctx, key)).hasValue(value);
        assertEquals(3, mockWebServer.getRequestCount());

        // Get from kubernetes
        configMapStore = new ConfigMapStore(kubernetesClient);
        assertThat(configMapStore.getSerializedState(ctx, key)).hasValue(value);
        assertEquals(4, mockWebServer.getRequestCount());

        // Remove the key
        configMapStore.removeSerializedState(ctx, key);
        assertEquals(4, mockWebServer.getRequestCount());

        configMapStore.flush(ctx);
        assertEquals(5, mockWebServer.getRequestCount());

        // get is null, so don't flush
        configMapStore.getSerializedState(ctx, key);
        assertEquals(5, mockWebServer.getRequestCount());
        configMapStore.getSerializedState(ctx, key);
        assertEquals(5, mockWebServer.getRequestCount());
        configMapStore.getSerializedState(ctx, key);
        assertEquals(5, mockWebServer.getRequestCount());
    }
```

Reply:
We may need to added a extra cache to store the state that last updated. When the flush is called, then compare them. And call kubernetes if they are not same.

And it's better to add a test to check the flush doesn't call kubernetes if state isn't updated.

Reply:
In this case the flush only calls Kubernetes because there is still data in the state store that wasn't removed. Maybe we should add a `clearAll` method to the state store interface and simply call that with the contract that subsequent `clearAll()` and `flush()` actions should be idempotent and ideally should not call Kubernetes again and again

Reply:
Thanks for adding the tests. You are right that the retrieval will always go to Kubernetes. The current implementation has some bugs with the bookkeeping to minimize the number of API calls. It does make more requests than necessary. 

I've rewrote the implementation to minimize the number of times we go to Kubernetes. We only retrieve once now on lookup. If the ConfigMap does not exist, we will only create the ConfigMap once we flush back any changes.